Add updated date to feed using modified. Fix #1933

karlcow opened the pull request.

justinmayer commented:
@karlcow: Looks like tests are failing with your proposed PR.

@ingwinlu / @avaris : Failing tests notwithstanding, any comments on the change?

Scheirle commented on pelican/writers.py:
```diff
@@ -55,7 +55,9 @@ def _add_item_to_the_feed(self, feed, item):
             pubdate=set_date_tzinfo(
-                item.modified if hasattr(item, 'modified') else item.date,
+                item.date, self.settings.get('TIMEZONE', None)),
+            pubdate=set_date_tzinfo(
```
Don't we want to set the `updateddate` here?

karlcow replied:
Oh! yoohoo. I was 100% I had changed this. Sorry about this.

Scheirle commented on pelican/writers.py:
```diff
@@ -55,7 +55,9 @@ def _add_item_to_the_feed(self, feed, item):
+                item.date, self.settings.get('TIMEZONE', None)),
+            pubdate=set_date_tzinfo(
+                item.modified if hasattr(item, 'modified') else None,
                 self.settings.get('TIMEZONE', None)))
```
`set_date_tzinfo(None, [...])` may throw exceptions or does not return `None`.

How about:
```
=set_date_tzinfo(item.modified, self.settings.get('TIMEZONE', None)) if hasattr(item, 'modified') else None
```

karlcow commented:
Thanks @Scheirle 

ok I have 3 failing tests. 

```
→ python -m unittest discover
...................................................................................sssssssssssssss.FFF..............................................................................
======================================================================
FAIL: test_basic_generation_works (pelican.tests.test_pelican.TestPelican)
----------------------------------------------------------------------

======================================================================
FAIL: test_custom_generation_works (pelican.tests.test_pelican.TestPelican)
----------------------------------------------------------------------
======================================================================
FAIL: test_custom_locale_generation_works (pelican.tests.test_pelican.TestPelican)
Test that generation with fr_FR.UTF-8 locale works
----------------------------------------------------------------------
```

The two first are because of the modification I did. That's normal, except it's not easy to find the original tests given that the assertionerror is on the diff and not on the file used for the test. :)

The third one is about locale. I can't figure out why it is happening. ah Maybe just because of the date again.

ok the origin file is `src/pelican/samples/content/super_article.rst`

```
:tags: foo, bar, foobar
:date: 2010-12-02 10:14
:modified: 2013-11-17 23:29
:category: yeah
:author: Alexis Métaireau
```

* And all the tests with dates are failing because of the expected output. 
* It seems also the published date is not added. So I might have missed something in the feedgenerator.py code.

I need to check this.

avaris commented:
@karlcow Any update that changes expected output needs to update the test output as well. Check the [docs](http://docs.getpelican.com/en/latest/contribute.html#running-the-test-suite) for details on how to do that.

karlcow commented:
```
→ LC_ALL=en_US.utf8 pelican -o pelican/tests/output/custom/ \
>     -s samples/pelican.conf.py samples/content/
ERROR: Skipping pages/jinja2_template.html: could not find information about 'NameError: title'
Done: Processed 12 articles, 1 draft, 3 pages and 1 hidden page in 0.65 seconds.

(pelican)13:10:26 ~/code/pelican/src/pelican
→ LC_ALL=fr_FR.utf8 pelican -o pelican/tests/output/custom_locale/ \
>     -s samples/pelican.conf_FR.py samples/content/
ERROR: Skipping pages/jinja2_template.html: could not find information about 'NameError: title'
Done: Processed 12 articles, 1 draft, 3 pages and 1 hidden page in 0.55 seconds.

(pelican)13:10:42 ~/code/pelican/src/pelican
→ LC_ALL=en_US.utf8 pelican -o pelican/tests/output/basic/ \
>     samples/content/
WARNING: LOCALE option doesn't contain a correct value
WARNING: Feeds generated without SITEURL set properly may not be valid
WARNING: No timezone information specified in the settings. Assuming your timezone is UTC for feed generation. Check http://docs.getpelican.com/en/latest/settings.html#timezone for more information
ERROR: Skipping ./another_super_article-fr.rst: could not find information about 'NameError: date'
ERROR: Skipping ./draft_article.rst: could not find information about 'NameError: date'
ERROR: Skipping pages/jinja2_template.html: could not find information about 'NameError: title'
WARNING: Unable to find `/pictures/Sushi.jpg`, skipping url replacement.
WARNING: Unable to find `/pictures/Sushi_Macro.jpg`, skipping url replacement.
WARNING: Unable to find `/pictures/Fat_Cat.jpg`, skipping url replacement.
Done: Processed 11 articles, 0 drafts, 3 pages and 1 hidden page in 0.43 seconds.
```

Then 

```
(pelican)13:11:24 ~/code/pelican/src/pelican
→ git status
On branch 1933/1
Your branch is up-to-date with 'origin/1933/1'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   pelican/tests/output/basic/feeds/alexis-metaireau.atom.xml
	modified:   pelican/tests/output/basic/feeds/alexis-metaireau.rss.xml
	modified:   pelican/tests/output/basic/feeds/all-en.atom.xml
	modified:   pelican/tests/output/basic/feeds/all.atom.xml
	modified:   pelican/tests/output/basic/feeds/yeah.atom.xml
	modified:   pelican/tests/output/custom/feeds/alexis-metaireau.atom.xml
	modified:   pelican/tests/output/custom/feeds/alexis-metaireau.rss.xml
	modified:   pelican/tests/output/custom/feeds/all-en.atom.xml
	modified:   pelican/tests/output/custom/feeds/all.atom.xml
	modified:   pelican/tests/output/custom/feeds/all.rss.xml
	modified:   pelican/tests/output/custom/feeds/yeah.atom.xml
	modified:   pelican/tests/output/custom/feeds/yeah.rss.xml
	modified:   pelican/tests/output/custom_locale/feeds/alexis-metaireau.atom.xml
	modified:   pelican/tests/output/custom_locale/feeds/alexis-metaireau.rss.xml
	modified:   pelican/tests/output/custom_locale/feeds/all-en.atom.xml
	modified:   pelican/tests/output/custom_locale/feeds/all.atom.xml
	modified:   pelican/tests/output/custom_locale/feeds/all.rss.xml
	modified:   pelican/tests/output/custom_locale/feeds/yeah.atom.xml
	modified:   pelican/tests/output/custom_locale/feeds/yeah.rss.xml

no changes added to commit (use "git add" and/or "git commit -a")
```

Then 

```
→ python -m unittest discover
...................................................................................sssssssssssssss..................................................................................
----------------------------------------------------------------------
Ran 180 tests in 5.090s

OK (skipped=15)
```

karlcow commented:
oh.
→ subl ~/.virtualenvs/pelican/lib/python2.7/site-packages/feedgenerator-1.7-py2.7.egg/feedgenerator/django/utils/feedgenerator.py

```
        item = {
            'title': to_unicode(title),
            'link': iri_to_uri(link),
            'description': to_unicode(description),
            'author_email': to_unicode(author_email),
            'author_name': to_unicode(author_name),
            'author_link': iri_to_uri(author_link),
            'pubdate': pubdate,
            'comments': to_unicode(comments),
            'unique_id': to_unicode(unique_id),
            'enclosure': enclosure,
            'categories': categories or (),
            'item_copyright': to_unicode(item_copyright),
            'ttl': ttl,
        }
```

while on 1.9 
https://docs.djangoproject.com/en/1.9/_modules/django/utils/feedgenerator/#SyndicationFeed.add_item

```
item = {
            'title': to_unicode(title),
            'link': iri_to_uri(link),
            'description': to_unicode(description),
            'author_email': to_unicode(author_email),
            'author_name': to_unicode(author_name),
            'author_link': iri_to_uri(author_link),
            'pubdate': pubdate,
            'updateddate': updateddate,
            'comments': to_unicode(comments),
            'unique_id': to_unicode(unique_id),
            'unique_id_is_permalink': unique_id_is_permalink,
            'enclosures': enclosures,
            'categories': categories or (),
            'item_copyright': to_unicode(item_copyright),
            'ttl': ttl,
        }
```

but if I do 

```
→ pip search feedgenerator
feedgenerator (1.7)  - Standalone version of django.utils.feedgenerator, compatible with Py3k
  INSTALLED: 1.7 (latest)
```

grmph.

karlcow commented:
Ah https://pypi.python.org/pypi/feedgenerator
Opening an issue there. 
https://github.com/dmdm/feedgenerator-py3k/issues/9

Scheirle commented:
@karlcow https://github.com/getpelican/feedgenerator is the current version on pypi.
The current dev version (master branch) also already supports `updateddate`.

So we only have to wait (*hint hint*) for the next release.

karlcow commented:
Ah sweet. ok. thanks.

karlcow commented:
Ah sweet. ok. thanks.

justinmayer commented:
@Scheirle / @karlcow: I just published [version 1.8](https://pypi.python.org/pypi/feedgenerator/1.8). Anything else I can do to help?

karlcow commented:
@justinmayer Wonderful!
@Scheirle Should I bump the version number in requirements.txt to add to this PR? :)
Also Discussion happening on #1938 about feedgenerator on pypi

karlcow commented:
OK. Tests are running smoothly and now the `published` and `updated` are available. 

```
→ python -m unittest discover
...................................................................................sssssssssssssss..................................................................................
----------------------------------------------------------------------
Ran 180 tests in 5.087s

OK (skipped=15)
```

I pushed the output too. If I need to remove them. Tell me. Maybe it's not necessary.

Thanks!

justinmayer commented:
Many thanks to both @karlcow and @Scheirle for working on this!

justinmayer commented:
I just ran `python -m unittest discover` and am seeing three failing tests on current master (d6ac93a), with at least one failure being seemingly related to `<updated>` tags being expected but absent in the output. This is on Mac OS 10.11.6 and Python 2.7.11. Anyone else seeing this?
